{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html('https://www.neb.com/tools-and-resources/selection-charts/alphabetized-list-of-recognition-specificities')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[                        Recognition Sequence               Enzyme\n",
       " 0                                    AA/CGTT                 AclI\n",
       " 1                                    A/AGCTT  HindIII-HF® HindIII\n",
       " 2                                    AAT/ATT        SspI-HF® SspI\n",
       " 3                                      /AATT                MluCI\n",
       " 4                                    A/CATGT                 PciI\n",
       " 5                                    A/CCGGT        AgeI-HF® AgeI\n",
       " 6                                ACCTGC(4/8)          BspMI BfuAI\n",
       " 7                                   A/CCWGGT                SexAI\n",
       " 8                                    A/CGCGT        MluI-HF® MluI\n",
       " 9                               ACGGC(12/14)                BceAI\n",
       " 10                                     A/CGT             HpyCH4IV\n",
       " 11                                    ACN/GT            HpyCH4III\n",
       " 12                  (10/15)ACNNNNGTAYC(12/7)                 BaeI\n",
       " 13                   (9/12)ACNNNNNCTCC(10/7)                BsaXI\n",
       " 14                                   A/CRYGT               AflIII\n",
       " 15                                   A/CTAGT        SpeI-HF® SpeI\n",
       " 16                               ACTGG(1/-1)                 BsrI\n",
       " 17                               ACTGGG(5/4)                 BmrI\n",
       " 18                                   A/GATCT                BglII\n",
       " 19                                   AGC/GCT                 AfeI\n",
       " 20                                     AG/CT                 AluI\n",
       " 21                                   AGG/CCT                 StuI\n",
       " 22                                   AGT/ACT             ScaI-HF®\n",
       " 23                                   AT/CGAT           BspDI ClaI\n",
       " 24   ATCTATGTCGGGTGCGGAGAAAGAGGTAAT(-15/-19)              PI-SceI\n",
       " 25                                   ATGCA/T        NsiI-HF® NsiI\n",
       " 26                                   AT/TAAT                 AseI\n",
       " 27                                 ATTT/AAAT                 SwaI\n",
       " 28                (11/13)CAANNNNNGTGG(12/10)                CspCI\n",
       " 29                                   C/AATTG        MfeI-HF® MfeI\n",
       " ..                                       ...                  ...\n",
       " 210                                  RGCGC/Y                HaeII\n",
       " 211                                    RG/CY              CviKI-1\n",
       " 212                                 RG/GNCCY             EcoO109I\n",
       " 213                                 RG/GWCCY                PpuMI\n",
       " 214      TAACTATAACGGTCCTAAGGTAGCGAA(-9/-13)               I-CeuI\n",
       " 215                                  TAC/GTA                SnaBI\n",
       " 216               TAGGGATAACAGGGTAAT(-9/-13)               I-SceI\n",
       " 217                                  T/CATGA                BspHI\n",
       " 218                                  T/CCGGA                BspEI\n",
       " 219                            TCCRAC(20/18)                 MmeI\n",
       " 220                                    T/CGA                TaqαI\n",
       " 221                                  TCG/CGA        NruI-HF® NruI\n",
       " 222                                   TCN/GA              Hpy188I\n",
       " 223                                  TC/NNGA            Hpy188III\n",
       " 224                                  T/CTAGA                 XbaI\n",
       " 225                                  T/GATCA         BclI-HF BclI\n",
       " 226                                    TG/CA              HpyCH4V\n",
       " 227                                  TGC/GCA                 FspI\n",
       " 228  TGGCAAACAGCTATTATGGGTATTATGGGT(-13/-17)              PI-PspI\n",
       " 229                                  TGG/CCA                 MscI\n",
       " 230                                  T/GTACA      BsrGI-HF® BsrGI\n",
       " 231                                    T/TAA                 MseI\n",
       " 232                                TTAAT/TAA                 PacI\n",
       " 233                                  TTA/TAA                 PsiI\n",
       " 234                                  TT/CGAA                BstBI\n",
       " 235                                  TTT/AAA                 DraI\n",
       " 236                                VC/TCGAGB                PspXI\n",
       " 237                                  W/CCGGW                BsaWI\n",
       " 238                                  YAC/GTR                BsaAI\n",
       " 239                                  Y/GGCCR                 EaeI\n",
       " \n",
       " [240 rows x 2 columns]]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.20.3'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Recognition Sequence</th>\n",
       "      <th>Enzyme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AA/CGTT</td>\n",
       "      <td>AclI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/AGCTT</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAT/ATT</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/AATT</td>\n",
       "      <td>MluCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/CATGT</td>\n",
       "      <td>PciI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Recognition Sequence               Enzyme\n",
       "0              AA/CGTT                 AclI\n",
       "1              A/AGCTT  HindIII-HF® HindIII\n",
       "2              AAT/ATT        SspI-HF® SspI\n",
       "3                /AATT                MluCI\n",
       "4              A/CATGT                 PciI"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dfs[0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[                            1\n",
       " 0                            \n",
       " A                     Adenine\n",
       " C                    Cytosine\n",
       " G                     Guanine\n",
       " T (or U)  Thymine (or Uracil)\n",
       " R                      A or G\n",
       " Y                      C or T\n",
       " S                      G or C\n",
       " W                      A or T\n",
       " K                      G or T\n",
       " M                      A or C\n",
       " B                 C or G or T\n",
       " D                 A or G or T\n",
       " H                 A or C or T\n",
       " V                 A or C or G\n",
       " N                    any base\n",
       " . or -                    gap,      1              2\n",
       " 0                    \n",
       " A  Ala        Alanine\n",
       " C  Cys       Cysteine\n",
       " D  Asp  Aspartic Acid\n",
       " E  Glu  Glutamic Acid\n",
       " F  Phe  Phenylalanine\n",
       " G  Gly        Glycine\n",
       " H  His      Histidine\n",
       " I  Ile     Isoleucine\n",
       " K  Lys         Lysine\n",
       " L  Leu        Leucine\n",
       " M  Met     Methionine\n",
       " N  Asn     Asparagine\n",
       " P  Pro        Proline\n",
       " Q  Gln      Glutamine\n",
       " R  Arg       Arginine\n",
       " S  Ser         Serine\n",
       " T  Thr      Threonine\n",
       " V  Val         Valine\n",
       " W  Trp     Tryptophan\n",
       " Y  Tyr       Tyrosine]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iupac_codes = pd.read_html('https://www.bioinformatics.org/sms/iupac.html', skiprows=1, index_col=0)\n",
    "iupac_codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>Adenine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>Cytosine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>Guanine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T (or U)</th>\n",
       "      <td>Thymine (or Uracil)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>A or G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>C or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>G or C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>A or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>A or C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>C or G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>A or G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>A or C or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>A or C or G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>any base</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>. or -</th>\n",
       "      <td>gap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            1\n",
       "0                            \n",
       "A                     Adenine\n",
       "C                    Cytosine\n",
       "G                     Guanine\n",
       "T (or U)  Thymine (or Uracil)\n",
       "R                      A or G\n",
       "Y                      C or T\n",
       "S                      G or C\n",
       "W                      A or T\n",
       "K                      G or T\n",
       "M                      A or C\n",
       "B                 C or G or T\n",
       "D                 A or G or T\n",
       "H                 A or C or T\n",
       "V                 A or C or G\n",
       "N                    any base\n",
       ". or -                    gap"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degenerate_nucleotides = iupac_codes[0]\n",
    "degenerate_nucleotides.columns \n",
    "degenerate_nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>Adenine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>Cytosine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>Guanine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>Thymine (or Uracil)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>A or G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>C or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>G or C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>A or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>A or C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>C or G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>A or G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>A or C or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>A or C or G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>any base</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>.</th>\n",
       "      <td>gap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     1\n",
       "0                     \n",
       "A              Adenine\n",
       "C             Cytosine\n",
       "G              Guanine\n",
       "T  Thymine (or Uracil)\n",
       "R               A or G\n",
       "Y               C or T\n",
       "S               G or C\n",
       "W               A or T\n",
       "K               G or T\n",
       "M               A or C\n",
       "B          C or G or T\n",
       "D          A or G or T\n",
       "H          A or C or T\n",
       "V          A or C or G\n",
       "N             any base\n",
       ".                  gap"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degenerate_nucleotides.index = degenerate_nucleotides.index.map(lambda x: x[0])\n",
    "degenerate_nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>A or G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>C or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>G or C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>A or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>A or C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>C or G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>A or G or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>A or C or T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>A or C or G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>any base</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             1\n",
       "0             \n",
       "R       A or G\n",
       "Y       C or T\n",
       "S       G or C\n",
       "W       A or T\n",
       "K       G or T\n",
       "M       A or C\n",
       "B  C or G or T\n",
       "D  A or G or T\n",
       "H  A or C or T\n",
       "V  A or C or G\n",
       "N     any base"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degenerate_nucleotides = degenerate_nucleotides.drop(['.', \"A\", \"C\", 'G', 'T'], errors='ignore')\n",
    "degenerate_nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>letters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>A or G</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>C or T</td>\n",
       "      <td>CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>G or C</td>\n",
       "      <td>GC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>A or T</td>\n",
       "      <td>AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>G or T</td>\n",
       "      <td>GT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>A or C</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>C or G or T</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>A or G or T</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>A or C or T</td>\n",
       "      <td>ACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>A or C or G</td>\n",
       "      <td>ACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>any base</td>\n",
       "      <td>ACGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             1 letters\n",
       "0                     \n",
       "R       A or G      AG\n",
       "Y       C or T      CT\n",
       "S       G or C      GC\n",
       "W       A or T      AT\n",
       "K       G or T      GT\n",
       "M       A or C      AC\n",
       "B  C or G or T     CGT\n",
       "D  A or G or T     AGT\n",
       "H  A or C or T     ACT\n",
       "V  A or C or G     ACG\n",
       "N     any base    ACGT"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degenerate_nucleotides['letters'] = degenerate_nucleotides[1].str.replace(' or ', '')\n",
    "degenerate_nucleotides.loc['N', 'letters'] = 'ACGT'\n",
    "degenerate_nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B': 'CGT',\n",
       " 'D': 'AGT',\n",
       " 'H': 'ACT',\n",
       " 'K': 'GT',\n",
       " 'M': 'AC',\n",
       " 'N': 'ACGT',\n",
       " 'R': 'AG',\n",
       " 'S': 'GC',\n",
       " 'V': 'ACG',\n",
       " 'W': 'AT',\n",
       " 'Y': 'CT'}"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degenerate_map = degenerate_nucleotides['letters'].to_dict()\n",
    "degenerate_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Recognition Sequence</th>\n",
       "      <th>Enzyme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AA/CGTT</td>\n",
       "      <td>AclI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/AGCTT</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAT/ATT</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/AATT</td>\n",
       "      <td>MluCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/CATGT</td>\n",
       "      <td>PciI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Recognition Sequence               Enzyme\n",
       "0              AA/CGTT                 AclI\n",
       "1              A/AGCTT  HindIII-HF® HindIII\n",
       "2              AAT/ATT        SspI-HF® SspI\n",
       "3                /AATT                MluCI\n",
       "4              A/CATGT                 PciI"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Recognition Sequence</th>\n",
       "      <th>Enzyme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACGTT</td>\n",
       "      <td>AclI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATATT</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATT</td>\n",
       "      <td>MluCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACATGT</td>\n",
       "      <td>PciI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACCGGT</td>\n",
       "      <td>AgeI-HF® AgeI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACCTGC</td>\n",
       "      <td>BspMI BfuAI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACCWGGT</td>\n",
       "      <td>SexAI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACGCGT</td>\n",
       "      <td>MluI-HF® MluI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACGGC</td>\n",
       "      <td>BceAI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Recognition Sequence               Enzyme\n",
       "0               AACGTT                 AclI\n",
       "1               AAGCTT  HindIII-HF® HindIII\n",
       "2               AATATT        SspI-HF® SspI\n",
       "3                 AATT                MluCI\n",
       "4               ACATGT                 PciI\n",
       "5               ACCGGT        AgeI-HF® AgeI\n",
       "6               ACCTGC          BspMI BfuAI\n",
       "7              ACCWGGT                SexAI\n",
       "8               ACGCGT        MluI-HF® MluI\n",
       "9                ACGGC                BceAI"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Recognition Sequence'] = df['Recognition Sequence'].str.replace('/', '')\n",
    "df['Recognition Sequence'] = df['Recognition Sequence'].str.strip('()/-1234567890')\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Recognition Sequence</th>\n",
       "      <th>Enzyme</th>\n",
       "      <th>non_degenerate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACGTT</td>\n",
       "      <td>AclI</td>\n",
       "      <td>AACGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "      <td>AAGCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATATT</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "      <td>AATATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATT</td>\n",
       "      <td>MluCI</td>\n",
       "      <td>AATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACATGT</td>\n",
       "      <td>PciI</td>\n",
       "      <td>ACATGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Recognition Sequence               Enzyme non_degenerate\n",
       "0               AACGTT                 AclI         AACGTT\n",
       "1               AAGCTT  HindIII-HF® HindIII         AAGCTT\n",
       "2               AATATT        SspI-HF® SspI         AATATT\n",
       "3                 AATT                MluCI           AATT\n",
       "4               ACATGT                 PciI         ACATGT"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['non_degenerate'] = df['Recognition Sequence']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Recognition Sequence</th>\n",
       "      <th>Enzyme</th>\n",
       "      <th>non_degenerate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACGTT</td>\n",
       "      <td>AclI</td>\n",
       "      <td>AACGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "      <td>AAGCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATATT</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "      <td>AATATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATT</td>\n",
       "      <td>MluCI</td>\n",
       "      <td>AATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACATGT</td>\n",
       "      <td>PciI</td>\n",
       "      <td>ACATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACCGGT</td>\n",
       "      <td>AgeI-HF® AgeI</td>\n",
       "      <td>ACCGGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACCTGC</td>\n",
       "      <td>BspMI BfuAI</td>\n",
       "      <td>ACCTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACC[AT]GGT</td>\n",
       "      <td>SexAI</td>\n",
       "      <td>ACCWGGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACGCGT</td>\n",
       "      <td>MluI-HF® MluI</td>\n",
       "      <td>ACGCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACGGC</td>\n",
       "      <td>BceAI</td>\n",
       "      <td>ACGGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Recognition Sequence               Enzyme non_degenerate\n",
       "0               AACGTT                 AclI         AACGTT\n",
       "1               AAGCTT  HindIII-HF® HindIII         AAGCTT\n",
       "2               AATATT        SspI-HF® SspI         AATATT\n",
       "3                 AATT                MluCI           AATT\n",
       "4               ACATGT                 PciI         ACATGT\n",
       "5               ACCGGT        AgeI-HF® AgeI         ACCGGT\n",
       "6               ACCTGC          BspMI BfuAI         ACCTGC\n",
       "7           ACC[AT]GGT                SexAI        ACCWGGT\n",
       "8               ACGCGT        MluI-HF® MluI         ACGCGT\n",
       "9                ACGGC                BceAI          ACGGC"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for degenerate_letter, replacement_letters in degenerate_map.items():\n",
    "    df['Recognition Sequence'] = df['Recognition Sequence'].str.replace(degenerate_letter, f'[{replacement_letters}]')\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus_name</th>\n",
       "      <th>enzyme</th>\n",
       "      <th>restriction_sequence</th>\n",
       "      <th>count</th>\n",
       "      <th>restriction_sequence_nondegenerate</th>\n",
       "      <th>restriction_sequence_nondegenerate_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>AclI</td>\n",
       "      <td>AACGTT</td>\n",
       "      <td>1</td>\n",
       "      <td>AACGTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>4</td>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "      <td>AATATT</td>\n",
       "      <td>2</td>\n",
       "      <td>AATATT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>MluCI</td>\n",
       "      <td>AATT</td>\n",
       "      <td>73</td>\n",
       "      <td>AATT</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>PciI</td>\n",
       "      <td>ACATGT</td>\n",
       "      <td>5</td>\n",
       "      <td>ACATGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    virus_name               enzyme restriction_sequence  \\\n",
       "0  gi|9626945|ref|NC_001498.1|                 AclI               AACGTT   \n",
       "1  gi|9626945|ref|NC_001498.1|  HindIII-HF® HindIII               AAGCTT   \n",
       "2  gi|9626945|ref|NC_001498.1|        SspI-HF® SspI               AATATT   \n",
       "3  gi|9626945|ref|NC_001498.1|                MluCI                 AATT   \n",
       "4  gi|9626945|ref|NC_001498.1|                 PciI               ACATGT   \n",
       "\n",
       "   count restriction_sequence_nondegenerate  \\\n",
       "0      1                             AACGTT   \n",
       "1      4                             AAGCTT   \n",
       "2      2                             AATATT   \n",
       "3     73                               AATT   \n",
       "4      5                             ACATGT   \n",
       "\n",
       "   restriction_sequence_nondegenerate_count  \n",
       "0                                         1  \n",
       "1                                         4  \n",
       "2                                         2  \n",
       "3                                        73  \n",
       "4                                         5  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import re \n",
    "\n",
    "records = SeqIO.parse('/Users/olgabot/Downloads/Mononegav_254genomes_sequence_100517.fasta', 'fasta')\n",
    "records\n",
    "\n",
    "\n",
    "lines = []\n",
    "\n",
    "for record in records:\n",
    "\n",
    "    for i, (sequence, enzyme, nondegenerate_sequence) in df.iterrows():\n",
    "        if not sequence in record:\n",
    "            pass\n",
    "        else:\n",
    "            nondegenerate_found = re.findall(nondegenerate_sequence, str(record.seq))\n",
    "            degenerate_found = re.findall(sequence, str(record.seq))\n",
    "            line = [record.id, enzyme, sequence, len(degenerate_found), nondegenerate_sequence, len(nondegenerate_found)]\n",
    "            lines.append(line)\n",
    "#             print(record.id, enzyme, sequence, len(found))\n",
    "\n",
    "found_restriction_sites = pd.DataFrame(\n",
    "    lines, columns=['virus_name', 'enzyme', 'restriction_sequence', 'count', \n",
    "                    'restriction_sequence_nondegenerate', 'restriction_sequence_nondegenerate_count'])  \n",
    "found_restriction_sites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus_name</th>\n",
       "      <th>enzyme</th>\n",
       "      <th>restriction_sequence</th>\n",
       "      <th>count</th>\n",
       "      <th>restriction_sequence_nondegenerate</th>\n",
       "      <th>restriction_sequence_nondegenerate_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>AclI</td>\n",
       "      <td>AACGTT</td>\n",
       "      <td>1</td>\n",
       "      <td>AACGTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>HindIII-HF® HindIII</td>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>4</td>\n",
       "      <td>AAGCTT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>SspI-HF® SspI</td>\n",
       "      <td>AATATT</td>\n",
       "      <td>2</td>\n",
       "      <td>AATATT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>MluCI</td>\n",
       "      <td>AATT</td>\n",
       "      <td>73</td>\n",
       "      <td>AATT</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>PciI</td>\n",
       "      <td>ACATGT</td>\n",
       "      <td>5</td>\n",
       "      <td>ACATGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>AgeI-HF® AgeI</td>\n",
       "      <td>ACCGGT</td>\n",
       "      <td>4</td>\n",
       "      <td>ACCGGT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>BspMI BfuAI</td>\n",
       "      <td>ACCTGC</td>\n",
       "      <td>8</td>\n",
       "      <td>ACCTGC</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>BceAI</td>\n",
       "      <td>ACGGC</td>\n",
       "      <td>9</td>\n",
       "      <td>ACGGC</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>HpyCH4IV</td>\n",
       "      <td>ACGT</td>\n",
       "      <td>14</td>\n",
       "      <td>ACGT</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gi|9626945|ref|NC_001498.1|</td>\n",
       "      <td>SpeI-HF® SpeI</td>\n",
       "      <td>ACTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>ACTAGT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    virus_name               enzyme restriction_sequence  \\\n",
       "0  gi|9626945|ref|NC_001498.1|                 AclI               AACGTT   \n",
       "1  gi|9626945|ref|NC_001498.1|  HindIII-HF® HindIII               AAGCTT   \n",
       "2  gi|9626945|ref|NC_001498.1|        SspI-HF® SspI               AATATT   \n",
       "3  gi|9626945|ref|NC_001498.1|                MluCI                 AATT   \n",
       "4  gi|9626945|ref|NC_001498.1|                 PciI               ACATGT   \n",
       "5  gi|9626945|ref|NC_001498.1|        AgeI-HF® AgeI               ACCGGT   \n",
       "6  gi|9626945|ref|NC_001498.1|          BspMI BfuAI               ACCTGC   \n",
       "7  gi|9626945|ref|NC_001498.1|                BceAI                ACGGC   \n",
       "8  gi|9626945|ref|NC_001498.1|             HpyCH4IV                 ACGT   \n",
       "9  gi|9626945|ref|NC_001498.1|        SpeI-HF® SpeI               ACTAGT   \n",
       "\n",
       "   count restriction_sequence_nondegenerate  \\\n",
       "0      1                             AACGTT   \n",
       "1      4                             AAGCTT   \n",
       "2      2                             AATATT   \n",
       "3     73                               AATT   \n",
       "4      5                             ACATGT   \n",
       "5      4                             ACCGGT   \n",
       "6      8                             ACCTGC   \n",
       "7      9                              ACGGC   \n",
       "8     14                               ACGT   \n",
       "9      1                             ACTAGT   \n",
       "\n",
       "   restriction_sequence_nondegenerate_count  \n",
       "0                                         1  \n",
       "1                                         4  \n",
       "2                                         2  \n",
       "3                                        73  \n",
       "4                                         5  \n",
       "5                                         4  \n",
       "6                                         8  \n",
       "7                                         9  \n",
       "8                                        14  \n",
       "9                                         1  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "found_restriction_sites.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "enzyme\n",
       "AscI             1\n",
       "SrfI             4\n",
       "NotI-HF® NotI    4\n",
       "AsiSI            5\n",
       "FseI             7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzyme_counts = found_restriction_sites.groupby('enzyme')['count'].sum()\n",
    "enzyme_counts.sort_values().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "enzyme\n",
       "AscI             1\n",
       "SrfI             4\n",
       "NotI-HF® NotI    4\n",
       "AsiSI            5\n",
       "FseI             7\n",
       "Name: restriction_sequence_nondegenerate_count, dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzyme_counts = found_restriction_sites.groupby('enzyme')['restriction_sequence_nondegenerate_count'].sum()\n",
    "enzyme_counts.sort_values().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
